### PR TITLE
Add  DEX  support for Cryptopower DCR wallets

### LINKED ISSUE #245: Make dcr wallet compatible with DCRDEX
This requires that `*dcr.Asset` implements: https://github.com/decred/dcrdex/blob/541b7d94b78e0188c3fbd1c89cbb86ffb02e04fd/client/asset/dcr/wallet.go#L68

EDIT: I'll work on this, @dreacot.

### PR DESCRIPTION
**Pull request by ukane-philemon**
This MR adds DEX support for cyptopower dcr wallets. An edge case I'm yet to handle is handling wallet removal when it is used by the DEX core. Closes #245.

- Bond payments ✅ 
- Trade: coming soon.

Builds on top of #238.

### CODE REVIEW

**Comment by dreacot** [comment]
@itswisdomagain please take a look

**Review by JoeGruffins — APPROVED** [review]
I'm able to set up a private server to test.

The first time I tried to post bond I saw this but it worked the second try:
image

If I leave the page while waiting on the bonds it forgot my custom server location and certificate. And now I can only see:
image


But I guess that's just where the development is at this point?

From the dex server, so I guess all went well.
```
2023-12-08 16:47:44.482 [DBG] AUTH: Activating new bond 4e6fc99f9f375516c1245e0da30096d45dfc01bbf56690384226fe73955e7c47:0 (dcr) committing 10000000 for user d1cf48bbfab73575be3214dfae02de6a9ccf35cee9d64eb8f046a904d01dbb25
2023-12-08 16:47:44.482 [INF] AUTH: Creating new user account d1cf48bbfab73575be3214dfae02de6a9ccf35cee9d64eb8f046a904d01dbb25 from 127.0.0.1, posted first bond in 4e6fc99f9f375516c1245e0da30096d45dfc01bbf56690384226fe73955e7c47:0 (dcr)
2023-12-08 16:47:44.488 [INF] AUTH: Bond accepted: acct d1cf48bbfab73575be3214dfae02de6a9ccf35cee9d64eb8f046a904d01dbb25 from 127.0.0.1 locked 10000000 in 4e6fc99f9f375516c1245e0da30096d45dfc01bbf56690384226fe73955e7c47:0. Bond total 2, tier 2
2023-12-08 16:47:44.497 [DBG] AUTH: User d1cf48bbfab73575be3214dfae02de6a9ccf35cee9d64eb8f046a904d01dbb25 score = 0:0 (0 successes) - 0 (violations) - 0 (0 preimage misses) 
2023-12-08 16:47:44.498 [INF] AUTH: Authenticated account d1cf48bbfab73575be3214dfae02de6a9ccf35cee9d64eb8f046a904d01dbb25 from 127.0.0.1 with 0 active orders, 0 active matches, tier = 2, bond tier = 2, score = 0
```

File: libwallet/assets/dcr/dex_asset.go [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+// AddressPrivKey fetches the privkey for the specified address.
+// Part of the Wallet interface.
+func (da *DEXAsset) AddressPrivKey(ctx context.Context, addr stdaddr.Address) (*secp256k1.PrivateKey, error) {
+	privKey, _, err := da.asset.Internal().DCR.LoadPrivateKey(ctx, addr)
+	return privKey, err
+}
+
+// Part of the Wallet interface.
+func (da *DEXAsset) Reconfigure(_ context.Context, cfg *dexasset.WalletConfig, _ dex.Network, _, _ string) (restart bool, err error) {
+	return cfg.Type != walletTypeSPV, nil
```
[the remark this answers is not in the document]
Comment (itswisdomagain) [reply libwallet/assets/dcr/dex_asset.go:510]: No need to implement the `LiveReconfigurer` interface.

File: dexc/core.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+		dexWallet, ok := wallet.(*dcrasset.Asset)
+		if !ok {
+			return nil, fmt.Errorf("DEX wallet not supported for %s", walletParams.Name)
+		}
+
+		return dcrasset.NewDEXAsset(dexWallet, uint32(accountNumber), accountName), nil
+	}
+
+	err := dcr.RegisterCustomWallet(walletMaker, def)
+	if err != nil && !strings.Contains(err.Error(), "already support" /* this is part of the error string returned if we've already registered a customer wallet */) {
```
[the remark this answers is not in the document]
Comment (ukane-philemon) [reply dexc/core.go:206]: Yes.

**Comment by ukane-philemon** [comment]
Thanks for the review @JoeGruffins. I’ll get to them.

**Comment by JustinBeBoy** [comment]
Seems like everything I commented in  #238 , it's hard to tell the changes in PR compared to PR238, I'll come back when PR238 is merged

File: libwallet/assets/dcr/dex_asset.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ukane-philemon) [reply libwallet/assets/dcr/dex_asset.go:510]: > No need to implement the `LiveReconfigurer` interface.

It's required by `dexdcr.Wallet` interface.

**Comment by ukane-philemon** [comment]
@JoeGruffins, your comments have been resolved. Please check to confirm.

Note to reviewers: The add server page does not remember custom server details if the user leaves without posting bonds.

**Comment by JoeGruffins** [comment]
Still seeing that not logged in popup with a fresh cryptopower and then it works the second time.

**Comment by ukane-philemon** [comment]
> Still seeing that not logged in popup with a fresh cryptopower and then it works the second time.

Huh 🤔.

File: libwallet/assets/dcr/dex_asset.go [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+	return nil, fmt.Errorf("unsupported address type %T", ka)
+}
+
+// AccountOwnsAddress checks if the provided address belongs to the
+// specified account.
+// Part of the Wallet interface.
+func (da *DEXAsset) AccountOwnsAddress(ctx context.Context, addr stdaddr.Address, acctName string) (bool, error) {
+	a, err := da.asset.Internal().DCR.KnownAddress(ctx, addr)
+	if err != nil {
+		return false, utils.TranslateError(err)
```
Comment (itswisdomagain) [inline libwallet/assets/dcr/dex_asset.go:120]: Upstream doesn't return an error for not exist:
```
if errors.Is(err, walleterrors.NotExist) {
	return false, nil
}
```

File: libwallet/assets/dcr/dex_asset.go [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// AccountBalance returns the balance breakdown for the specified account.
+// Part of the Wallet interface.
+func (da *DEXAsset) AccountBalance(_ context.Context, confirms int32, _ string) (*walletjson.GetAccountBalanceResult, error) {
+	bal, err := da.asset.GetAccountBalance(int32(da.dexAccountNumber), &confirms)
```
Comment (itswisdomagain) [inline libwallet/assets/dcr/dex_asset.go:137]: I see why `asset.GetAccountBalance` was modified to accept confs now. Please use the internal wallet method for getting account balance. And use the provided ctx as well.

File: libwallet/assets/dcr/dex_asset.go [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+		Unconfirmed:             bal.UnConfirmed.ToCoin(),
+		VotingAuthority:         bal.VotingAuthority.ToCoin(),
+	}, nil
+}
+
+// LockedOutputs fetches locked outputs for the Wallet.
+// Part of the Wallet interface.
+func (da *DEXAsset) LockedOutputs(ctx context.Context, _ string) ([]chainjson.TransactionInput, error) {
+	return da.asset.Internal().DCR.LockedOutpoints(ctx, da.dexAccountName)
+}
```
Comment (itswisdomagain) [inline libwallet/assets/dcr/dex_asset.go:158]: When mixing is enabled, multiple accounts may be in use. Always use the acctName parameter passed to methods.

File: libwallet/assets/dcr/dex_asset.go [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+	"github.com/decred/dcrd/wire"
+)
+
+const (
+	walletTypeSPV = "SPV"
+)
+
+// DEXAsset wraps *Asset and implements dexdcr.Wallet.
+type DEXAsset struct {
+	asset *Asset
```
Comment (itswisdomagain) [inline libwallet/assets/dcr/dex_asset.go:41]: Change this to `wallet *wallet.Wallet`. Use direct methods of the upstream dcrwallet object. That should make this file look more closely like the upstream dex spv file.

**Review by itswisdomagain — DISMISSED** [review]
Will review again after these.

**Review by dreacot — CHANGES_REQUESTED** [review]
after posting bonds, i'm pretty sure i was shown to pay  fee of 0.2 dcr, but i only got debited 0.1 dcr

also i believe a minimum of 1 confirmation is required before you can proceed, but on leaving the trade page and returning to it, i get taken to the market page, i shouldn't go past the bond page until my payment is confirmed
Screenshot from 2023-12-13 10-33-04

Screenshot from 2023-12-13 10-33-25

why does this keep showing every time i go to the dex page
Screenshot from 2023-12-13 10-33-18

**Comment by ukane-philemon** [comment]
> also i believe a minimum of 1 confirmation is required before you can proceed, but on leaving the trade page and returning to it, i get taken to the market page, i shouldn't go past the bond page until my payment is confirmed

Yes, but once an exchange has been added to core I don't think we need to cache the bond data somewhere just to make the user wait on the bond confirmation page. At this point the bond payment has already been made.

**Comment by ukane-philemon** [comment]
> i'm pretty sure i was shown to pay fee of 0.2 dcr

Yes, I was previously accounting for bond maintenance.

**Comment by JoeGruffins** [comment]
I see insufficient balance now. The testnet server also seems to be up now.


https://github.com/crypto-power/cryptopower/assets/34998433/d8243057-9ba5-4039-9cc5-2e78df48e045

**Comment by ukane-philemon** [comment]
> I see insufficient balance now.

Hmm, I've tested several times (with block wait times over 13min)  yet I didn't see this. I'm still looking and will suggest some changes.

**Comment by ukane-philemon** [comment]
> > I see insufficient balance now.
> 
> Hmm, I've tested several times (with block wait times over 13min) yet I didn't see this. I'm still looking and will suggest some changes.

@JoeGruffins, this was caused by a "nasty bug". `v0.6.3/client/asset/dcr/config.go#L33` required specifying a primary account which has been removed from dex master (for spv wallets only). Fixed by [3c7246f](https://github.com/crypto-power/cryptopower/pull/269/commits/3c7246f58cb9f2c112252e0a7daa31dde3e7c3e0).

**Review by dreacot — APPROVED** [review]
Tested ACK

**Comment by ukane-philemon** [comment]
@JoeGruffins, please can you confirm the error has been fixed on your end?

**Review by JoeGruffins — APPROVED** [review]
No problems now!

**Review by JustinBeBoy — APPROVED** [review]

**Pull request merged**